### DISCUSSION TITLE
Feedback for the experimental `advance-branches` feature

### DISCUSSION THREAD

**Discussion by emesterhazy** [jj-vcs/jj#3549] [General]
Please leave any feedback that you have for the experimental `advance-branches` config setting added in 

- https://github.com/martinvonz/jj/pull/3129

This feature should be available in the next (0.17) release or today if you build from source at head.

------------------

## Update from ilyagr:

Here is a sample configuration from the PR description linked above:

```toml
[experimental-advance-branches]
enabled-branches = ["glob:*"]
disabled-branches = ["main", "glob:push-*"]
```

If/when there are more docs, it might be nice to link them here.

**Comment by emesterhazy** [comment]
There were a few open questions from a [discussion on the PR](https://github.com/martinvonz/jj/pull/3129#discussion_r1571229550).

1. Should we do nothing if there are multiple branches eligible to advance?
2. Should we only advance a branch if there isn't already a branch pointing to the target commit?
3. Should we add a special case for new commits with multiple parents (aka merges) at moves a branch to the merge commit?

Feel free to share your thoughts on these as well.

**Comment by Cretezy** [comment]
Are docs going to be ready for the 0.17 release? I could take up writing or reviewing them.

**Reply by emesterhazy** [reply]
I wasn't planning to include this in the release notes since it's an experimental feature, but perhaps it's not a bad idea. There might be some additional polish we want to do first though, such as the `split` issue raised below.

**Comment by BatmanAoD** [comment]
One other issue I've noticed: `split` should probably move the branch forward, so that it's still at `@-` after the operation is done.

**Reply by bnjmnt4n** [reply]
Related issue: https://github.com/martinvonz/jj/issues/3419.

**Comment by tp-woven** [comment]
Thanks for implementing this @emesterhazy!

Just wanted to chime in on the `@` vs `@-` discussion:

> Whether or not to advance to @ or @- doesn't actually make much difference in the code. I can add another config option to control that behavior.

I am very much in favor of advancing to `@`. This is because my workflow usually looks like this:

1. `jj co main -m"working on whatever"` (or sometimes without the message, and I run `jj describe` later)
2. Change some code
3. `jj git push --change=@`
4. Iterate 2+3 until ready for review
5. Get PR comments
6. `jj new -m"address pr comments"`
7. `jj branch set <branch name>` <- ideally, this shouldn't be needed
8. Change some more code
9. `jj git push` <- pushes `@`
10. Iterate 8+9 until ready for merge

I guess most folks use the `squash` workflow and continually push their changes down, in which case `@-` makes sense, but personally I prefer this (mostly to make it easier to see diffs in IDEs, etc.)

EDIT: Added steps 4 and 10 for clarification

**Reply by BatmanAoD** [reply]
I also create chains of commits, and I've [previously explained](https://github.com/martinvonz/jj/issues/2338#issuecomment-1951626994) that I don't think updating branches to `@` makes sense. This isn't a difference between `squash` versus chains of commits; it's a difference between considering the working copy to be more of a "scratchpad" or something to be published.

Note that with the current implementation, if you skip step 6, after you've changed your code in step 7, `jj new` and `jj commit` will both prepare your branch for `jj git push`. So your workflow is supported, you're just missing a step.

In general, I think `jj` encourages keeping `@` one ahead of your "committed" work. All of these commands create a new, empty `@` commit:

* `new` and its aliases:
  * `commit`
  * `checkout`
* `backout` (though the placement of the empty commit [is weird and will probably change](https://github.com/martinvonz/jj/issues/2802))

...with that said, I do actually see the logic of having `@` be pushed in some circumstances, which is why I [originally proposed](https://github.com/martinvonz/jj/issues/2338#issuecomment-1938072659) making `jj describe` be the command implementing auto-advancement. But I think that ship has probably sailed.

**Reply by BatmanAoD** [reply]
There was more discussion about `@` vs `@-` [here](https://github.com/martinvonz/jj/pull/3129#issuecomment-1964796020).

**Reply by tp-woven** [reply]
Thanks for the references and the explanation @BatmanAoD!

I think my previous post didn't communicate this very well (at all? 😅), but in most cases my workflow doesn't actually involve "multiple" commits, just two (one for the main change, one for comments). Basically, in the workflow above, I iterate over 2+3 multiple times, and then after review I iterate over 7+8 (8+9 in the "edited" workflow). (This is the reason I compared it to the "squash" workflow, as that's the alternative for "iterating" on the same commit when the branch is set to `@-`.)

**Reply by BatmanAoD** [reply]
If you're only ever updating the branch once per feature, why do you need automatic branch advancement in the first place?

**Reply by tp-woven** [reply]
I don't think it's about how many times you do it per feature, it's more that it's easy to forget to create/move the branch.

**Reply by BatmanAoD** [reply]
This feature certainly can't help with forgetting to create the branch, of course.

Why not just adopt a squash-based workflow, rather than advocating for this feature (which is already arguably supporting a niche use case, per those prior discussions) to work differently?

**Reply by BatmanAoD** [reply]
To put it differently, you are effectively using a squash-like workflow, in that you are repeatedly modifying an existing commit. From the perspective of the final commit history and the sequence of commits seen on the remote, there's no difference between what you're doing and repeatedly `squash`-ing. Why not just embrace that?

**Reply by tp-woven** [reply]
> This feature certainly can't help with forgetting to create the branch, of course.

No, but the branch is already created by `jj git push --change=@`. This feature _can_ help with forgetting to update it.

> Why not just adopt a squash-based workflow

The main reasons I work on the commit itself rather than squashing are shell and IDE integrations - I prefer to see the status/diffs of the "whole" commit while I work, rather than "what I changed since the last squash".

**Reply by BatmanAoD** [reply]
Ah. This draft MR would resolve that, I think: https://github.com/martinvonz/jj/pull/3405

**Reply by tp-woven** [reply]
Oh, nice! That could be really useful!

**Comment by amiryal** [comment]
Thanks @emesterhazy, it’s a very welcome ergonomic improvement!

> * Should we do nothing if there are multiple branches eligible to advance?

Advance all of them.

> * Should we only advance a branch if there isn't already a branch pointing to the target commit?

For `jj commit`: this sounds like a situation that can only happen in the edit workflow, for which the use case of `advance-branches` is not clear to me to begin with. If anything, then the branch pointing to the target (old `@`) should advance to the **new** `@`.

For `jj new`: forgive me, but I do not see the use case for `advance-branches` at all. Can we add a configuration to limit this feature to `jj commit` only?

> * Should we add a special case for new commits with multiple parents (aka merges) at moves a branch to the merge commit?

Advance from the first parent only, maybe? Tough question; it could be another configuration toggle.

**Reply by BatmanAoD** [reply]
The open questions were actually the result of a discussion about how to handle merges. In the current implementation, there's no good way to merge two "advanceable" branches and only advance one of them.

The "only advance a branch if there isn't already a branch pointing to the target commit" idea was suggested so that users could explicitly select which branch to advance:

```bash
jj new <branch1> <branch2>
jj branch set <branch1> -r @
jj branch commit -m "merge <branch2> into <branch1>"
```

> For `jj new`: forgive me, but I do not see the use case for advance-branches at all. Can we add a configuration to limit this feature to `jj commit` only?

`jj commit` is effectively just `jj new`, except that `-m` (if provided) targets the _old_ commit rather than the _new_ commit. I don't know if it's possible or desirable to further differentiate their behavior.

> Advance from the first parent only, maybe?

If I understand correctly, there's not really a "first parent" in the JJ commit metadata, but yes, advancing the first listed parent listed in the `new` command itself is the proposal in question. (Originally suggested [here](https://github.com/martinvonz/jj/pull/3129#discussion_r1545499345).)

**Comment by allancalix** [comment]
This is a great quality of life feature thanks for adding this! One thing I want to note as I set this is I wanted to have some sensible global defaults (only advance on branch with some pr glob convention) and disable on main.

I then set `jj config set --repo experimental-advance-branches.enabled-branches '["main"]'` on a repo where I push directly on top of main but the `enabled-branches` in the local config and `disabled-branches` in the global config conflicted. It's not that obvious how these config options get merged together and will behave as a result.

I just ended up setting both options locally but a couple ideas come to mind:
- log a warning if matching branches are in both enabled and disabled sets
- enforce in the config option that both options be set at the same level

Either way I think the flexibility here is awesome, especially with the glob support.

Here's what I ended up with for reference:
[Shell function to override per repo](https://github.com/allancalix/dotfiles/blob/f42f7b7e3eb96320fe8e439b9666258f933032bc/nix/home.nix#L194)
[Global options preventing this behavior on main](https://github.com/allancalix/dotfiles/blob/f42f7b7e3eb96320fe8e439b9666258f933032bc/nix/home.nix#L417)

**Reply by BatmanAoD** [reply]
Intuitively, I'd expect it to work as follows:

Start with globally enabled branches, subtract the globally excluded branches, add the locally enabled branches (ignoring global exclusions in case of conflict), subtract locally disabled branches.

Or, equivalently (and probably closer to an actual implementation), for any given branch name:

* Match against local exclusions. If there's a match, stop; don't advance. 
* Match against local inclusions. If there's a match, stop; do advance. 
* Match against global exclusions. If there's a match, stop; don't advance.
* Match against global inclusions. If there's a match, advance.
* Otherwise, don't advance.

Of course this could be extended to more than two levels of configuration.

**Comment by amiryal** [comment]
Are there any thoughts about extending this feature to `jj rebase`? If some branches point to `@-` before the operation, and that commit gets rebased away from `@-`, then any advance-eligible branches should remain at the new `@-` (instead of being moved to the rebased commit). This behaviour would be especially handy when shuffling the order of commits within the same line of work with `jj rebase -r x -B y`, always keeping the branch pointing to the last commit in the chain.

**Reply by emesterhazy** [reply]
That's an interesting suggestion. I think the core of the issue is that branches in `jj` don't really work the same way that branches work in Git, and I'm a bit apprehensive about trying to make them work the same way in all or most cases. If we add too many differences we'll end up with two fundamentally different branch models, and I'm not sure that's a good idea.

I wonder how this same issue would look with topics (https://github.com/martinvonz/jj/issues/3402). Maybe it wouldn't be a problem because all of the commits would be tagged for the same topic before and after the rebase and the push algorithm would figure out where to set the branch for git compatibility.

**Reply by amiryal** [reply]
I was not aware of the Topics effort, thanks! I agree, branches as they exist today in jj are really just bookmarks (and not so good ones at that), and their main power today is for Git interoperability.

So, when thinking about how auto-advancing branches should behave, one needs to keep in mind the Topics discussion, because I also agree that it would make sense for the topic bookmark→git branch to “remain where it should be”, relative to the topic.

**Comment by NoahTheDuke** [comment]
I read the above and didn't see this, so apologies if it was already covered.

I've tried this out, and the one issue I have is that if I have a branch `b1` off `main`, and I have config `enabled-branches = ["*"]`, when I run `jj commit -m "new change"` on `b1`, it advances `main` as well, no matter how far back `main` is. This makes it mostly useless for me as I make all of my new changes off `main` but I want to advance `main` only when I commit directly to it or when I merge.

I'd prefer if the functionality only advanced the first branch found by looking backwards along the commit graph. That way, I can have `aaa (main) --> bbb (b1) --> ccc (HEAD)`, and then I type `jj commit` and now it's `aaa (main) --> bbbb -> ccc (b1) --> ddd (HEAD)`. If it's `aaa --> bbb (main) (b1) --> ccc (HEAD)` and I type in `jj commit`, then moving both makes sense.

**Reply by BatmanAoD** [reply]
I actually thought this _did_ only look one commit back, but I hadn't tested that. I'm not sure going back more than one commit is wrong, though.

To prevent auto-advancing trunk branches, you should configure `disabled-branches = ["main", "master"]`.

**Reply by emesterhazy** [reply]
It should only look one commit back and advance the branch at `@-` to `@` before the new commit is created. I'm not sure exactly what is happening in your case, but it sounds like either a misconception about how the feature works, or a bug in the implementation.

Can you replicate the issue and copy and paste the output of the jj commands to show the graph before and after the commit?

**Comment by ilyagr** [comment]
I am trying this out for a while, and it's been mostly good so far. However, I just have had one bad experience with `experimental-advance-branches`, though in a pretty messy situation. I think I have to disable it :(

**Setup:** I have three branches, `forget-local`, `bug3223`, and `changeid` that I'm working on, and a merge of them called `ilya2` that I use to compile my personal version of `jj`. Whether this is a thing of beauty or horror, I leave it up to the reader:

![image](https://github.com/martinvonz/jj/assets/4123047/ffa6dcfd-9500-431b-93b2-caab2c8d0ab6)

My config is:

```toml
[experimental-advance-branches]
enabled-branches = ["glob:*"]
disabled-branches = ["main", "gh-pages"]
```

**Problem:** Now, I `jj new ilya2` without any intention of moving branches. However, all the branches move confusingly:

![image](https://github.com/martinvonz/jj/assets/4123047/2a35e9bc-050c-4ead-988a-5a0e32efb1e3)

I'm guessing this is the experimental-advance-branches's fault, but it doesn't seem essential to the "advancing branches" functionality, so I hope it's fixable.


**Update:** I thought that maybe adding `ilya2` to `disabled-branches` would help (this would make sense for my setup regardless of whether it helps with this problem), but it did not help with this problem.

**Reply by BatmanAoD** [reply]
This is the exact reason for this call-out in the [top comment](https://github.com/martinvonz/jj/discussions/3549#discussioncomment-9174663):

> Should we only advance a branch if there isn't already a branch pointing to the target commit?

This question came out of a [long discussion](https://github.com/martinvonz/jj/pull/3129#discussion_r1536559897) in the original MR. I think that advancing a branch to a commit that already has a branch is pretty clearly the wrong behavior, so I [tried implementing a check to prevent that](https://github.com/martinvonz/jj/pull/3879/files), but it doesn't work yet because `self.repo().view().local_branches_for_commit(id)` takes a _change_ ID, not a _commit_ ID. I haven't had time to go back and figure out the right logic yet.

**Reply by ilyagr** [reply]
Good to know, thanks! I haven't followed the discussion carefully, I'm just reporting this as a user.

**Reply by amiryal** [reply]
I already [said before](https://github.com/martinvonz/jj/discussions/3549#discussioncomment-9493575):
> For `jj new`: forgive me, but I do not see the use case for `advance-branches` at all. Can we add a configuration to limit this feature to `jj commit` only?

The reply to that was:
> `jj commit` is effectively just `jj new`, except that `-m` (if provided) targets the _old_ commit rather than the _new_ commit. I don't know if it's possible or desirable to further differentiate their behavior.

To which I did not respond at the time, but now we are witnessing a clear case in support of my argument: `jj commit` is not just a convenience shortcut — it actually conveys a difference of intent! If the target is transitioning from being “unfinished” to being “finished”, then I _would_ like branches to advance to it, but otherwise _not_.

Here is how I suggest to identify target commits for advancing branches without further differentiating between `commit` and `new`. Instead of “when creating a new commit on top”, change it to “when describing a previously uncommented non-merge commit”. WDYT?

**Reply by BatmanAoD** [reply]
Sorry to link to yet another long discussion with extra context, but it was suggested [here](https://github.com/martinvonz/jj/issues/2338#issuecomment-1944749176) that, since `commit` is simply `new` + `describe`, it could eventually be deprecated. But I actually agree that the semantics for "advance this branch" correspond exclusively to _finishing_ work rather than _starting_ work, and that these are separate; in my [original proposal](https://github.com/martinvonz/jj/issues/2338#issuecomment-1938072659), therefore, I suggested making this behavior part of `describe` rather than `new`. Either way, since `commit` incorporates the behavior of `describe`, it would have the advance-branch behavior.

I think the heuristic of only advancing branches to "previously uncommented non-merge commits" is unnecessarily restrictive and somewhat arbitrary, though; I suspect it would cause confusion.

**Reply by spoerri** [reply]
Another approach would be to advance branches when branches are read instead of written.

So advance branches on `jj new someBranch` and `jj git push`. And that'd be whether or not the branch has a description!

But this is just regarding the `jj commit` talk. I don't understand what that has to do with ilyagr's issue.

**Reply by BatmanAoD** [reply]
What do you mean by "read instead of written"? `new` changes local state, and `push` changes remote state.

**Reply by spoerri** [reply]
Yeah, that's why I chose those two examples, because they're so different.

"Read instead of written" means: advance a branch to its latest commit whenever that branch value is actually used, instead of when it would be triggered to change by a write like `jj new/commit` or `jj describe`.

Just new/commit would be sad because it'd be entrenching the need for a child commit to consider something "done", and describe is kind of weird and surprising. If we did rely on new/commit, we'd probably also want to trigger on squash. And I can't even imagine how `jj edit` would be expected to work with advance branches on write. Checking that the branch is up-to-date every time it is consulted doesn't have any of those issues.

So, of course you need to update the local representation of a branch before you push to its remote.

I've been on a small crusade against the jj data model ever depending on a final empty commit, and this may be the only real application.

**Reply by BatmanAoD** [reply]
Okay, so it would be on "read" operations _and_ "write" operations?

I honestly think that changing the data model on read operations is surprising, in a bad way.

My understanding is that the "final empty commit" is pretty well baked into `jj` at this point, as evidenced by `checkout` being a deprecated alias for `new`. I could be mistaken, though.

**Reply by spoerri** [reply]
It could be not so much read vs write as: **all** jj commands should automatically update all jj's state based on the working copy.

That's how the "edit" command works, right?

**Reply by BatmanAoD** [reply]
Maybe I'm misunderstanding, or you wrote something different from what you meant, but no, `edit` updates the working copy based on jj's (and potentially git's) internal state; never the reverse.

**Reply by spoerri** [reply]
When using `jj edit`, even `jj status` causes the working copy to be “applied”, to rebase subsequent commits.

**Reply by BatmanAoD** [reply]
Oh, I see, you're saying that after setting the working copy to be an existing commit using `jj edit`, snapshotting the working copy (with any `jj` command) will cause the descendants to be rebased.

I'm still not sure how this is analogous to advancing _branches_. The branches are only ever part of the backend internal state, not the files in the working copy.

**Reply by spoerri** [reply]
> after setting the working copy to be an existing commit using jj edit, snapshotting the working copy (with any jj command) will cause the descendants to be rebased.

That's right.

> not sure how this is analogous to advancing branches

1. We have a precedent for updating _some_ repo state automatically from the working copy, triggered by any jj command. This is actually _more_ compelling to do for advancing branches than for rebasing! The whole character of branch advancing is automatic. Rebasing is more surprising, so if we're happy to rebase automatically on any jj command, we should certainly be happy to do it for advancing branches.
2. Advancing branches should work for the actual `jj edit` case too. There's no justification for jj to rebase descendants but leave branches pointing to stale commits (aside for implementation priority of course)

To be clear: this would obviate `jj commit`, and can cleanly support a clean workflow:

    jj new -m "cool feature"
    ... hack hack ...
    jj git push -r @     # automatically advance branches to @ first

And that makes "-r @" an attractive default for `jj git push`, just like for many jj commands.

**Reply by BatmanAoD** [reply]
I guess I'm just fundamentally not understanding what problem you think this would solve. So far most of the issues have been from people seeing branches be advanced that they didn't want to advance. If the branch automatically advanced _while pushing_, there wouldn't be any opportunity to even check that what you're pushing is what you want to push.

**Comment by BatmanAoD** [comment]
One issue that has come up a couple times for me now, and seems like it might be a bug: if you have a bookmark that is one commit ahead of another, and you use `new` to "switch" from the one behind to the one ahead, it updates the branch that's behind, which doesn't seem correct to me.

Demo:

```
❯ jj bookmark create stay-behind
Created 1 bookmarks pointing to yrsssxum b0f767e1 stay-behind | (empty) (no description set)

❯ jj commit -m"stay-behind commit"
Working copy now at: syyxzlyp a6b46978 (empty) (no description set)
Parent commit      : yrsssxum 5ff6a440 stay-behind | (empty) stay-behind commit

❯ jj bookmark create one-ahead
Created 1 bookmarks pointing to syyxzlyp a6b46978 one-ahead | (empty) (no description set)

❯ jj describe -m"commit that is one ahead of stay-behind"
Working copy now at: syyxzlyp 9f1956a4 one-ahead | (empty) commit that is one ahead of stay-behind
Parent commit      : yrsssxum 5ff6a440 stay-behind | (empty) stay-behind commit

❯ jj new stay-behind
Working copy now at: pxsonutv 97372339 (empty) (no description set)
Parent commit      : yrsssxum 5ff6a440 stay-behind | (empty) stay-behind commit

❯ jj new one-ahead
Working copy now at: wqyqkuyx f8ef382e (empty) (no description set)
Parent commit      : syyxzlyp 9f1956a4 one-ahead stay-behind | (empty) commit that is one ahead of stay-behind
```

Note that after `jj new one-ahead`, `stay-behind` is now on the same commit as `one-ahead`.

**Comment by nilskch** [comment]
@emesterhazy I would have expected that the new change/commit has the bookmark of the previous commit. 
<img width="649" alt="Screenshot 2025-05-26 at 13 38 26" src="https://github.com/user-attachments/assets/1156505d-b3db-42cd-8dd8-41ff07ba9c6f" />

I assume you are using `jj commit` all the time and never `jj new` ? :D This feels more natural with `jj commit`, but it does not feel natural with `jj new` in my opinion.

The tests assume that `jj desc` + `jj new` == `jj commit`, but I think this assumption is wrong here. I want to always push the `@` commit, not the one before.

Your design leads to unexpected behaviour to people like me who use `jj new`:

<img width="687" alt="before" src="https://github.com/user-attachments/assets/5fe90e7b-41c9-4a0c-823e-055050d99bc1" />

What do you think?

**Reply by bnjmnt4n** [reply]
See https://github.com/jj-vcs/jj/pull/3129#issuecomment-1964796020.

**Comment by mfarrugi** [comment]
This might be a bit of a meta-comment, but it took me a while to figure out what I was missing, then finding out this secret feature existed. This could be somewhere in the docs aimed at someone trying out jujutsu starting from a `jj git init --colocate` in a real repository with multiple existing branches.

I'm coming from a git workflow where I have a few medium-lived branches that aren't related, occasionally with separate upstream branches (eg. main, release, and shared/feature), and many short-lived branches. I regularly rebase each branch on its upstream tracking branch, which requires some slow and complicated scripting that I'm hoping to remove by using Jujutsu instead. 

One thing that feels off to me so far is how to replace the notion of 'what branch am I on' and 'what branch am I targeting'. Bookmarking @ directly does not persist, it slides back to @- on commits, so that's not a way to emulate git branches with this (and I see whether @ or @- represents a branch is already a point of discussion). My next thought was to emulate 'current branch' and 'upstream branch' as `heads(::@ & bookmarks())` and `heads(::@ & remote_bookmarks())` but this doesn't line up when we're ahead of multiple bookmarks, and upstream cannot actually be inferred this way very often. The former seems like it would usually be 1 local bookmark anyway, but is there a concept that replaces git upstream branches? This seems like it's more of a problem for importing existing (git?) remote branches, since local bookmarks would all be created implicitly following some ancestor commit (ie. master@origin).

**Comment by nomeata** [comment]
>  In the current implementation, there's no good way to merge two "advanceable" branches and only advance one of them.

I just ran into this: I merged another branch into my branch and didn’t even notice it updated the other bookmark, only when I did `jj git push` I noticed that, and had to clean up.

Intuitively I’d be happy if the auto-advancing only ever goes from `first_parent(@)` to `@`, but maybe that’s still influenced by `git`-think and doesn't match how people use `jj`.
